fulltext/extract.py

```diff
@@ -1,57 +1,210 @@
+        ))
+        # Dispatch the extraction task.
+        extract.apply_async((identifier, id_type, version), {'token': token},
+                            task_id=_task_id)
+        logger.info('extract: started processing as %s', _task_id)
     except Exception as e:
-        logger.error('Failed to process %s: %s' % (document_id, e))
-        raise e
-    return {
-        'document_id': document_id,
-    }
+        logger.debug(e)
+        raise TaskCreationFailed('Failed to create task: %s', e) from e
+    return _task_id
+
+
+def get_task(identifier: str, id_type: str, version: str) -> Extraction:
+    """
+    Get the status of an extraction task.
+
+    Parameters
+    ----------
+    identifier : str
+        Unique identifier for the paper being extracted. Usually an arXiv ID.
+    id_type : str
+        Either 'arxiv' or 'submission'.
+    version : str
+        Extractor version.
+
+    Returns
+    -------
+    :class:`Extraction`
+
+    """
+    _task_id = task_id(identifier, id_type, version)
+    result = extract.AsyncResult(_task_id)
+    exception: Optional[str] = None
+    owner: Optional[str] = None
+    if result.status == 'PENDING':
+        raise NoSuchTask('No such task')
+    elif result.status in ['SENT', 'STARTED', 'RETRY']:
+        _status = Extraction.Status.IN_PROGRESS
+    elif result.status == 'FAILURE':
+        _status = Extraction.Status.FAILED
+        exception = str(result.result)
+    elif result.status == 'SUCCESS':
+        _status = Extraction.Status.SUCCEEDED
+        owner = str(result.result['owner'])
+    else:
+        raise RuntimeError(f'Unexpected state: {result.status}')
+    return Extraction(
+        identifier=identifier,
+        bucket=id_type,
+        task_id=_task_id,
+        version=version,
+        status=_status,
+        exception=exception,
+        owner=owner
+    )
+
 
+@celery_app.task
+def extract(identifier: str, id_type: str, version: str,
+            owner: Optional[str] = None,
+            token: Optional[str] = None) -> Dict[str, str]:
+    """Perform text extraction for a single arXiv document."""
+    logger.debug('Perform extraction for %s in bucket %s with version %s',
+                 identifier, id_type, version)
 
-extract_fulltext.async_result = AsyncResult
+    canonical = pdf.CanonicalPDF.current_session()
+    storage = store.Storage.current_session()
+    compilations = compiler.Compiler.current_session()
+
+    try:
+        if id_type == SupportedBuckets.ARXIV:
+            pdf_path = canonical.retrieve(identifier)
+        elif id_type == SupportedBuckets.SUBMISSION:
+            pdf_path, owner = compilations.retrieve(identifier, token)
+        else:
+            RuntimeError(f'Unsupported identifier: {identifier} ({id_type})')
+        extraction = storage.retrieve(identifier, version, bucket=id_type,
+                                      meta_only=True)
+        content = extractor.do_extraction(pdf_path)
+        assert content is not None
```

Comment:
I would expect this to be more like ```if content is None: logger.error(...etc...```

Why does this happen to be an assertion?

Reply:
Looks like that was left over from an earlier version. ``extractor.do_extraction`` (which is ``Extractor.__call__``) only returns a ``str``. Fixed in e0f4e24